File: apiendpoint/node/models/qcModels.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

/**
 * QC Inspection Plans Model
 * Master data for QC procedures and inspection requirements
 */
const QCInspectionPlan = sequelize.define('QCInspectionPlan', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    plan_code: {
        type: DataTypes.STRING(50),
        allowNull: false,
        unique: true,
        comment: 'QC Plan reference code'
    },
    part_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        references: {
            model: 'master_prod',
            key: 'part_number'
        }
    },
    plan_name: {
        type: DataTypes.STRING(255),
        allowNull: false
    },
    inspection_type: {
        type: DataTypes.ENUM('incoming', 'in_process', 'final', 'customer_return'),
        allowNull: false
    },
    sampling_method: {
        type: DataTypes.ENUM('100_percent', 'statistical', 'mil_std', 'custom'),
        defaultValue: 'statistical'
    },
    sample_size: {
        type: DataTypes.INTEGER,
        defaultValue: 1
    },
    acceptance_criteria: {
        type: DataTypes.JSON,
        comment: 'Criteria for pass/fail decisions'
    },
    inspection_points: {
        type: DataTypes.JSON,
        comment: 'List of measurement points and tolerances'
    },
    required_tools: {
        type: DataTypes.STRING(500),
        comment: 'Required inspection tools/equipment'
    },
    estimated_time_minutes: {
        type: DataTypes.INTEGER,
        defaultValue: 30
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    },
    created_by: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: 'users',
            key: 'id'
        }
    }
}, {
    tableName: 'qc_inspection_plans',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    comment: 'QC inspection plans and procedures'
});

/**
 * QC Inspection Results Model
 * Detailed inspection records and results
 */
const QCInspectionResult = sequelize.define('QCInspectionResult', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    inspection_number: {
        type: DataTypes.STRING(50),
        allowNull: false,
        unique: true,
        comment: 'QC inspection reference'
    },
    qc_plan_id: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: 'qc_inspection_plans',
            key: 'id'
        }
    },
    source_type: {
        type: DataTypes.ENUM('production', 'receiving', 'return', 'audit'),
        allowNull: false
    },
    source_reference_id: {
        type: DataTypes.INTEGER,
        comment: 'Reference to production order, delivery, etc'
    },
    lot_number: {
        type: DataTypes.STRING(255),
        allowNull: false
    },
    part_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        references: {
            model: 'master_prod',
            key: 'part_number'
        }
    },
    quantity_inspected: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false
    },
    quantity_passed: {
        type: DataTypes.DECIMAL(12, 3),
        defaultValue: 0.00
    },
    quantity_failed: {
        type: DataTypes.DECIMAL(12, 3),
        defaultValue: 0.00
    },
    quantity_rework: {
        type: DataTypes.DECIMAL(12, 3),
        defaultValue: 0.00
    },
    inspection_status: {
        type: DataTypes.ENUM('pending', 'in_progress', 'completed', 'on_hold'),
        defaultValue: 'pending'
    },
    overall_result: {
        type: DataTypes.ENUM('pass', 'fail', 'conditional_pass', 'pending'),
        defaultValue: 'pending'
    },
    inspector_id: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: 'users',
            key: 'id'
        }
    },
    inspection_start_time: {
        type: DataTypes.DATE,
        allowNull: true
    },
    inspection_end_time: {
        type: DataTypes.DATE,
        allowNull: true
    },
    inspection_location: {
        type: DataTypes.STRING(100),
        comment: 'Where inspection was performed'
    },
    measurement_data: {
        type: DataTypes.JSON,
        comment: 'Detailed measurement results'
    },
    defect_codes: {
        type: DataTypes.JSON,
        comment: 'List of defect codes found'
    },
    corrective_actions: {
        type: DataTypes.TEXT,
        comment: 'Required corrective actions'
    },
    inspector_notes: {
        type: DataTypes.TEXT,
        comment: 'Inspector comments and observations'
    },
    approved_by: {
        type: DataTypes.INTEGER,
        allowNull: true,
        references: {
            model: 'users',
            key: 'id'
        },
        comment: 'QC supervisor approval'
    },
    approved_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    approval_notes: {
        type: DataTypes.TEXT
    }
}, {
    tableName: 'qc_inspection_results',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    comment: 'Detailed QC inspection results'
});

/**
 * QC Defect Codes Model
 * Master data for defect classification
 */
const QCDefectCode = sequelize.define('QCDefectCode', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    defect_code: {
        type: DataTypes.STRING(20),
        allowNull: false,
        unique: true
    },
    defect_name: {
        type: DataTypes.STRING(255),
        allowNull: false
    },
    defect_category: {
        type: DataTypes.ENUM('dimensional', 'visual', 'functional', 'material', 'assembly', 'packaging'),
        allowNull: false
    },
    severity_level: {
        type: DataTypes.ENUM('critical', 'major', 'minor', 'observation'),
        allowNull: false
    },
    description: {
        type: DataTypes.TEXT
    },
    corrective_action_template: {
        type: DataTypes.TEXT,
        comment: 'Standard corrective action'
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    }
}, {
    tableName: 'qc_defect_codes',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: false,
    comment: 'Master defect codes for QC classification'
});

/**
 * QC Non-Conformance Model
 * Non-conformance reports for quality issues
 */
const QCNonConformance = sequelize.define('QCNonConformance', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    ncr_number: {
        type: DataTypes.STRING(50),
        allowNull: false,
        unique: true,
        comment: 'NCR reference number'
    },
    inspection_result_id: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: 'qc_inspection_results',
            key: 'id'
        }
    },
    ncr_type: {
        type: DataTypes.ENUM('supplier', 'internal', 'customer'),
        allowNull: false
    },
    part_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        references: {
            model: 'master_prod',
            key: 'part_number'
        }
    },
    lot_number: {
        type: DataTypes.STRING(255),
        allowNull: false
    },
    quantity_affected: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false
    },
    problem_description: {
        type: DataTypes.TEXT,
        allowNull: false
    },
    root_cause_analysis: {
        type: DataTypes.TEXT
    },
    immediate_action: {
        type: DataTypes.TEXT,
        comment: 'Immediate containment action'
    },
    corrective_action: {
        type: DataTypes.TEXT,
        comment: 'Long-term corrective action'
    },
    preventive_action: {
        type: DataTypes.TEXT,
        comment: 'Preventive measures'
    },
    ncr_status: {
        type: DataTypes.ENUM('open', 'investigating', 'action_required', 'verification', 'closed'),
        defaultValue: 'open'
    },
    priority: {
        type: DataTypes.ENUM('low', 'medium', 'high', 'critical'),
        defaultValue: 'medium'
    },
    reported_by: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: 'users',
            key: 'id'
        }
    },
    assigned_to: {
        type: DataTypes.INTEGER,
        allowNull: true,
        references: {
            model: 'users',
            key: 'id'
        },
        comment: 'Responsible person for resolution'
    },
    verified_by: {
        type: DataTypes.INTEGER,
        allowNull: true,
        references: {
            model: 'users',
            key: 'id'
        },
        comment: 'QC verification'
    },
    target_close_date: {
        type: DataTypes.DATEONLY,
        allowNull: true
    },
    actual_close_date: {
        type: DataTypes.DATEONLY,
        allowNull: true
    },
    cost_impact: {
        type: DataTypes.DECIMAL(12, 2),
        defaultValue: 0.00,
        comment: 'Financial impact'
    }
}, {
    tableName: 'qc_non_conformance',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    comment: 'Non-conformance reports for quality issues'
});

/**
 * Enhanced OQC Model
 * Updated Outgoing Quality Control model
 */
const OQC = sequelize.define('OQC', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    inspection_number: {
        type: DataTypes.STRING(50),
        allowNull: false,
        unique: true,
        comment: 'OQC reference number'
    },
    part_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        references: {
            model: 'master_prod',
            key: 'part_number'
        }
    },
    lot_number: {
        type: DataTypes.STRING(255),
        allowNull: false
    },
    production_order_id: {
        type: DataTypes.INTEGER,
        allowNull: true,
        references: {
            model: 'production_orders',
            key: 'id'
        },
        comment: 'Source production order'
    },
    quantity_received: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false,
        comment: 'Total quantity received for inspection'
    },
    quantity_inspected: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false,
        comment: 'Quantity actually inspected'
    },
    quantity_good: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false,
        defaultValue: 0.00
    },
    quantity_ng: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false,
        defaultValue: 0.00
    },
    quantity_rework: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false,
        defaultValue: 0.00
    },
    inspection_type: {
        type: DataTypes.ENUM('first_piece', 'in_process', 'final', 'random'),
        defaultValue: 'final'
    },
    inspection_date: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW
    },
    inspector_id: {
        type: DataTypes.INTEGER,
        allowNull: false,
        references: {
            model: 'users',
            key: 'id'
        }
    },
    inspection_location: {
        type: DataTypes.STRING(100),
        defaultValue: 'OQC-AREA'
    },
    overall_result: {
        type: DataTypes.ENUM('pass', 'fail', 'conditional_pass', 'pending'),
        defaultValue: 'pending'
    },
    inspection_status: {
        type: DataTypes.ENUM('pending', 'in_progress', 'completed', 'on_hold'),
        defaultValue: 'pending'
    },
    measurement_data: {
        type: DataTypes.JSON,
        comment: 'Inspection measurements and results'
    },
    defect_details: {
        type: DataTypes.JSON,
        comment: 'Details of any defects found'
    },
    inspector_notes: {
        type: DataTypes.TEXT
    },
    approved_by: {
        type: DataTypes.INTEGER,
        allowNull: true,
        references: {
            model: 'users',
            key: 'id'
        }
    },
    approved_at: {
        type: DataTypes.DATE,
        allowNull: true
    },
    approval_notes: {
        type: DataTypes.TEXT
    },
    disposition: {
        type: DataTypes.ENUM('release', 'hold', 'rework', 'scrap', 'return_to_supplier'),
        allowNull: true
    },
    disposition_notes: {
        type: DataTypes.TEXT
    }
}, {
    tableName: 'oqc',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    comment: 'Enhanced Outgoing Quality Control inspections'
});

// Define associations
QCInspectionPlan.associate = function(models) {
    QCInspectionPlan.belongsTo(models.User, { foreignKey: 'created_by', as: 'creator' });
    QCInspectionPlan.hasMany(QCInspectionResult, { foreignKey: 'qc_plan_id', as: 'inspection_results' });
};

QCInspectionResult.associate = function(models) {
    QCInspectionResult.belongsTo(QCInspectionPlan, { foreignKey: 'qc_plan_id', as: 'inspection_plan' });
    QCInspectionResult.belongsTo(models.User, { foreignKey: 'inspector_id', as: 'inspector' });
    QCInspectionResult.belongsTo(models.User, { foreignKey: 'approved_by', as: 'approver' });
    QCInspectionResult.hasMany(QCNonConformance, { foreignKey: 'inspection_result_id', as: 'non_conformances' });
};

QCNonConformance.associate = function(models) {
    QCNonConformance.belongsTo(QCInspectionResult, { foreignKey: 'inspection_result_id', as: 'inspection_result' });
    QCNonConformance.belongsTo(models.User, { foreignKey: 'reported_by', as: 'reporter' });
    QCNonConformance.belongsTo(models.User, { foreignKey: 'assigned_to', as: 'assignee' });
    QCNonConformance.belongsTo(models.User, { foreignKey: 'verified_by', as: 'verifier' });
};

OQC.associate = function(models) {
    OQC.belongsTo(models.User, { foreignKey: 'inspector_id', as: 'inspector' });
    OQC.belongsTo(models.User, { foreignKey: 'approved_by', as: 'approver' });
    OQC.belongsTo(models.ProductionOrder, { foreignKey: 'production_order_id', as: 'production_order' });
};

module.exports = {
    QCInspectionPlan,
    QCInspectionResult,
    QCDefectCode,
    QCNonConformance,
    OQC
};

File: apiendpoint/app/models/qc_models.py
"""
Quality Control (QC) Models for FastAPI
Enhanced QC system with comprehensive tracking
"""
from sqlalchemy import Column, Integer, String, Text, TIMESTAMP, DECIMAL, Boolean, DateTime, ForeignKey, JSON, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database.base import Base

class QCInspectionPlan(Base):
    __tablename__ = "qc_inspection_plans"
    
    id = Column(Integer, primary_key=True, index=True)
    plan_code = Column(String(50), unique=True, nullable=False, comment="QC Plan reference code")
    part_number = Column(String(100), ForeignKey("master_prod.part_number"), nullable=False)
    plan_name = Column(String(255), nullable=False)
    inspection_type = Column(Enum('incoming', 'in_process', 'final', 'customer_return', name='inspection_type_enum'), nullable=False)
    sampling_method = Column(Enum('100_percent', 'statistical', 'mil_std', 'custom', name='sampling_method_enum'), default='statistical')
    sample_size = Column(Integer, default=1)
    acceptance_criteria = Column(JSON, comment="Criteria for pass/fail decisions")
    inspection_points = Column(JSON, comment="List of measurement points and tolerances")
    required_tools = Column(String(500), comment="Required inspection tools/equipment")
    estimated_time_minutes = Column(Integer, default=30)
    is_active = Column(Boolean, default=True)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(TIMESTAMP, default=func.current_timestamp())
    updated_at = Column(TIMESTAMP, default=func.current_timestamp(), onupdate=func.current_timestamp())

class QCInspectionResult(Base):
    __tablename__ = "qc_inspection_results"
    
    id = Column(Integer, primary_key=True, index=True)
    inspection_number = Column(String(50), unique=True, nullable=False, comment="QC inspection reference")
    qc_plan_id = Column(Integer, ForeignKey("qc_inspection_plans.id"), nullable=False)
    source_type = Column(Enum('production', 'receiving', 'return', 'audit', name='source_type_enum'), nullable=False)
    source_reference_id = Column(Integer, comment="Reference to production order, delivery, etc")
    lot_number = Column(String(255), nullable=False)
    part_number = Column(String(100), ForeignKey("master_prod.part_number"), nullable=False)
    quantity_inspected = Column(DECIMAL(12,3), nullable=False)
    quantity_passed = Column(DECIMAL(12,3), default=0.00)
    quantity_failed = Column(DECIMAL(12,3), default=0.00)
    quantity_rework = Column(DECIMAL(12,3), default=0.00)
    inspector_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    inspection_start_time = Column(DateTime)
    inspection_end_time = Column(DateTime)
    inspection_location = Column(String(100))
    measurement_data = Column(JSON, comment="Actual measurements and values")
    defect_codes = Column(JSON, comment="List of defect codes found")
    corrective_actions = Column(Text, comment="Actions taken to address issues")
    inspector_notes = Column(Text)
    inspection_status = Column(Enum('pending', 'in_progress', 'completed', 'cancelled', name='inspection_status_enum'), default='pending')
    overall_result = Column(Enum('pass', 'fail', 'conditional', name='result_enum'))
    created_at = Column(TIMESTAMP, default=func.current_timestamp())
    updated_at = Column(TIMESTAMP, default=func.current_timestamp(), onupdate=func.current_timestamp())

class QCDefectCode(Base):
    __tablename__ = "qc_defect_codes"
    
    id = Column(Integer, primary_key=True, index=True)
    defect_code = Column(String(20), unique=True, nullable=False)
    defect_category = Column(String(50), nullable=False)
    defect_description = Column(String(255), nullable=False)
    severity_level = Column(Enum('critical', 'major', 'minor', 'cosmetic', name='severity_enum'), default='minor')
    standard_action = Column(Text, comment="Standard corrective action")
    is_active = Column(Boolean, default=True)
    created_at = Column(TIMESTAMP, default=func.current_timestamp())
    updated_at = Column(TIMESTAMP, default=func.current_timestamp(), onupdate=func.current_timestamp())

class QCNonConformance(Base):
    __tablename__ = "qc_non_conformance"
    
    id = Column(Integer, primary_key=True, index=True)
    ncr_number = Column(String(50), unique=True, nullable=False, comment="Non-conformance report number")
    inspection_result_id = Column(Integer, ForeignKey("qc_inspection_results.id"), nullable=False)
    ncr_type = Column(Enum('material', 'process', 'equipment', 'documentation', 'customer_complaint', name='ncr_type_enum'), nullable=False)
    part_number = Column(String(100), ForeignKey("master_prod.part_number"), nullable=False)
    lot_number = Column(String(255), nullable=False)
    quantity_affected = Column(DECIMAL(12,3), nullable=False)
    problem_description = Column(Text, nullable=False)
    immediate_action = Column(Text, nullable=False)
    priority = Column(Enum('low', 'medium', 'high', 'critical', name='priority_enum'), default='medium')
    status = Column(Enum('open', 'investigating', 'action_required', 'closed', 'cancelled', name='ncr_status_enum'), default='open')
    reported_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    assigned_to = Column(Integer, ForeignKey("users.id"))
    target_close_date = Column(DateTime)
    actual_close_date = Column(DateTime)
    created_at = Column(TIMESTAMP, default=func.current_timestamp())
    updated_at = Column(TIMESTAMP, default=func.current_timestamp(), onupdate=func.current_timestamp())

class QCCalibrationRecord(Base):
    __tablename__ = "qc_calibration_records"
    
    id = Column(Integer, primary_key=True, index=True)
    equipment_id = Column(String(50), nullable=False, comment="Equipment identification")
    equipment_name = Column(String(255), nullable=False)
    calibration_date = Column(DateTime, nullable=False)
    next_calibration_date = Column(DateTime, nullable=False)
    calibration_standard = Column(String(100), comment="Standard used for calibration")
    performed_by = Column(Integer, ForeignKey("users.id"), nullable=False)
    calibration_certificate = Column(String(255), comment="Certificate reference")
    calibration_status = Column(Enum('in_service', 'out_of_service', 'limited_use', name='calibration_status_enum'), default='in_service')
    notes = Column(Text)
    created_at = Column(TIMESTAMP, default=func.current_timestamp())
    updated_at = Column(TIMESTAMP, default=func.current_timestamp(), onupdate=func.current_timestamp())
